File: api/read-pdf.ts
import type { VercelRequest, VercelResponse } from "@vercel/node";
import pdfParse from "pdf-parse";

interface RentEntry {
  period: string;
  annualBaseRent?: string;
  monthlyInstallment?: string;
  perSquareFoot?: string;
}

function extractRentAmounts(text: string): RentEntry[] {
  const dollarPattern = /\$([\d,]+\.\d{2})/g;

  // Isolate the Base Rent section
  const rentSectionMatch = text.match(
    /Base\s*Rent[:\s]*(.*?)(?:\([l-z]\)|Additional\s+Rent)/is
  );
  if (!rentSectionMatch) return [];

  const rentText = rentSectionMatch[1];

  // Find all period labels
  const periodPatterns = [
    /(?:Rent\s+Commencement\s*(?:Date)?[-\s]*Lease\s+Years?\s*\d+)/gi,
    /Lease\s+Years?\s*\d+\s*[-–]\s*\d+/gi,
  ];

  const periodsMap = new Map<string, number>();
  for (const pp of periodPatterns) {
    let m: RegExpExecArray | null;
    while ((m = pp.exec(rentText)) !== null) {
      const label = m[0].trim();
      if (!periodsMap.has(label)) {
        periodsMap.set(label, m.index);
      }
    }
  }
  // Also check full text
  for (const pp of periodPatterns) {
    pp.lastIndex = 0;
    let m: RegExpExecArray | null;
    while ((m = pp.exec(text)) !== null) {
      const label = m[0].trim();
      if (!periodsMap.has(label)) {
        periodsMap.set(label, m.index);
      }
    }
  }

  const periodLabels = [...periodsMap.entries()]
    .sort((a, b) => a[1] - b[1])
    .map(([label]) => label);

  // Extract all dollar amounts from the rent section
  const amounts: number[] = [];
  let dm: RegExpExecArray | null;
  while ((dm = dollarPattern.exec(rentText)) !== null) {
    amounts.push(parseFloat(dm[1].replace(/,/g, "")));
  }

  if (!amounts.length || !periodLabels.length) return [];

  // Classify by magnitude
  const annual = amounts.filter((a) => a >= 10_000).sort((a, b) => a - b);
  const monthly = amounts
    .filter((a) => a >= 1_000 && a < 10_000)
    .sort((a, b) => a - b);
  const perSqft = amounts.filter((a) => a < 1_000).sort((a, b) => a - b);

  const fmt = (n: number) =>
    `$${n.toLocaleString("en-US", { minimumFractionDigits: 2, maximumFractionDigits: 2 })}`;

  return periodLabels.map((period, i) => {
    const entry: RentEntry = { period };
    if (i < annual.length) entry.annualBaseRent = fmt(annual[i]);
    if (i < monthly.length) entry.monthlyInstallment = fmt(monthly[i]);
    if (i < perSqft.length) entry.perSquareFoot = fmt(perSqft[i]);
    return entry;
  });
}

async function fetchBlob(blobUrl: string): Promise<Buffer> {
  const url = new URL(blobUrl);
  if (!url.hostname.endsWith(".vercel-storage.com")) {
    throw new Error("Invalid blob URL. Must be a Vercel Blob URL.");
  }
  const resp = await fetch(blobUrl);
  if (!resp.ok) {
    throw new Error(`Failed to fetch blob: ${resp.status} ${resp.statusText}`);
  }
  return Buffer.from(await resp.arrayBuffer());
}

export default async function handler(req: VercelRequest, res: VercelResponse) {
  if (req.method !== "POST") {
    res.setHeader("Allow", "POST");
    return res.status(405).json({ error: "Method not allowed. Use POST." });
  }

  const { blobUrl } = req.body as { blobUrl?: string };
  if (!blobUrl) {
    return res.status(400).json({
      error: "JSON body must include 'blobUrl'. Upload the file first via /api/upload-pdf.",
    });
  }

  let fileData: Buffer;
  let fileName: string;
  try {
    fileData = await fetchBlob(blobUrl);
    fileName = new URL(blobUrl).pathname.split("/").pop() || "document.pdf";
  } catch (err) {
    const message = err instanceof Error ? err.message : "Unknown error";
    return res.status(502).json({ error: message });
  }

  if (!fileName.toLowerCase().endsWith(".pdf")) {
    return res.status(400).json({ error: "Only PDF files are accepted." });
  }

  try {
    const pdfData = await pdfParse(fileData);
    const rentAmounts = extractRentAmounts(pdfData.text);

    return res.status(200).json({
      fileName,
      totalPages: pdfData.numpages,
      rentAmounts,
    });
  } catch (err) {
    console.error("PDF parse error:", err);
    const message = err instanceof Error ? err.message : "Unknown error";
    return res.status(500).json({ error: message });
  }
}

File: api/read_pdf.py
import json
import io
from http.server import BaseHTTPRequestHandler
from PyPDF2 import PdfReader


class handler(BaseHTTPRequestHandler):
    def do_POST(self):
        content_type = self.headers.get("Content-Type", "")

        if "multipart/form-data" not in content_type:
            self._send_json(400, {"error": "Content-Type must be multipart/form-data"})
            return

        # Parse boundary from content type
        boundary = None
        for part in content_type.split(";"):
            part = part.strip()
            if part.startswith("boundary="):
                boundary = part[len("boundary="):]
                break

        if not boundary:
            self._send_json(400, {"error": "No boundary found in Content-Type"})
            return

        content_length = int(self.headers.get("Content-Length", 0))
        body = self.rfile.read(content_length)

        # Parse the multipart data
        file_data, file_name = self._parse_multipart(body, boundary)

        if file_data is None:
            self._send_json(400, {"error": "No file uploaded"})
            return

        if not file_name.lower().endswith(".pdf"):
            self._send_json(400, {"error": "Only PDF files are accepted."})
            return

        try:
            reader = PdfReader(io.BytesIO(file_data))
            total_pages = len(reader.pages)
            text = "\n".join(page.extract_text() or "" for page in reader.pages)

            self._send_json(200, {
                "fileName": file_name,
                "totalPages": total_pages,
                "text": text,
            })
        except Exception as e:
            self._send_json(500, {"error": str(e)})

    def do_GET(self):
        self._send_json(405, {"error": "Method not allowed. Use POST."})

    def _parse_multipart(self, body, boundary):
        boundary_bytes = f"--{boundary}".encode()
        parts = body.split(boundary_bytes)

        for part in parts:
            if b"Content-Disposition" not in part:
                continue

            header_end = part.find(b"\r\n\r\n")
            if header_end == -1:
                continue

            headers_raw = part[:header_end].decode("utf-8", errors="replace")
            file_data = part[header_end + 4:]

            # Strip trailing \r\n-- 
            if file_data.endswith(b"\r\n"):
                file_data = file_data[:-2]
            if file_data.endswith(b"--"):
                file_data = file_data[:-2]
            if file_data.endswith(b"\r\n"):
                file_data = file_data[:-2]

            # Extract filename
            file_name = None
            for line in headers_raw.split("\r\n"):
                if "filename=" in line:
                    start = line.index('filename="') + len('filename="')
                    end = line.index('"', start)
                    file_name = line[start:end]
                    break

            if file_name:
                return file_data, file_name

        return None, None

    def _send_json(self, status, data):
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.end_headers()
        self.wfile.write(json.dumps(data).encode())
